maps/mp/isnipe/extras/anticamp.gsc

```diff
@@ -1,299 +1,196 @@
+    }
 
-if (level.campers[3] == self.name)
-{
-level.campers[3] = "";
-level.camperstimeleft[3] = "";
+    // Set settings to default when they are not set.
+    _setSettingIfNotExists("interfaceTime", 5);
+    _setSettingIfNotExists("killTime", 5);
+    _setSettingIfNotExists("minDistance", 100);
+    level.antiCamp["campers"] = [];
 }
 
-if (level.campers[4] == self.name)
-{
-level.campers[4] = "";
-level.camperstimeleft[4] = "";
+/**
+ * Set defaultValue in level.antiCamp[property] if it does not exist.
+ *
+ * @param property: the setting property to act upon
+ * @param defaultValue: the default value to set for the property if not set
+ */
+_setSettingIfNotExists(property, defaultValue) {
+    if (!isDefined(level.antiCamp[property])) {
+        level.antiCamp[property] = defaultValue;
+    }
 }
 
-if (level.campers[5] == self.name)
-{
-level.campers[5] = "";
-level.camperstimeleft[5] = "";
+/**
+ * Initialize the anti camp for an individual player.
+ */
+onPlayerConnected() {
+    player = self;
+    player.antiCamp = [];
+    player.antiCamp["hudElements"] = [];
+    player.antiCamp["time"] = 0;
+    player thread _createHUD();
+    player thread _antiCamp();
 }
 
+/**
+ * Create the anticamp HUD of the player.
+ *
+ * This function takes care of both drawing while the user is alive and
+ * deleting the HUD whenever the user dies.
+ */
+_createHUD() {
+    self endon("disconnect");
+    for(;;) {
+        player thread _drawHUD();
+        self waittill("death");
+        player thread _deleteHUD();
+        self waittill("spawned_player");
+    }
 }
 
-EnableAntiCamp(waittime)
-{
-    self endon("disconnect");
+/**
+ * Draw the anticamp HUD of the player.
+ *
+ * With the current implementation, the HUD is updated every 0.2 seconds.
+ */
+_drawHUD() {
     self endon("death");
+    self endon("disconnect");
 
-if (IsSubStr(self.name, "boaet")) //Disable anticamp for bots
-    {
+    killTime = level.antiCamp["interfaceTime"] + level.antiCamp["killTime"];
+    header = _createHUDRow(1.0, -25);
+    self.antiCamp["hudElements"][0] = header;
+    while(;;) {
+        campers = level.antiCamp["campers"];
+        hudElements = self.antiCamp["hudElements"];
+
+        // Hide HUD when no campers are available.
+        if (campers.size == 0) {
+            header setText("");
+        } else {
+            header setText("ANTICAMP");
+        }
+
+        // Update HUD for all available campers, and add rows if necessary.
+        for (i = 0; i < campers; i++) {
+            if (!isDefined(hudElements[i + 1])) {
+                hudElements[i + 1] = _createHUDRow(0.9, -15 + 10 * i);
+            }
+            hudElements[i + 1] setText(campers[i].name + " ^3(" + killTime - campers[i].antiCamp["time"] + ")");
+        }
+
+        // Remove unnecessary HUD elements because of lack of campers.
+        while (hudElements.size > campers.size + 1) {
+            // TODO: Check whether destroying elements in array removes them from array.
```

Comment:
But setting it to 'undefined' does delete it from the array (and therefore also the size property is adjusted)?

Reply:
Yes, the size will change, however the index of latter ones will not change.

The "arrays" in gsc are not really arrays/lists in other languages, they are more like dictionaries from Python or hashmaps from Java.
Sorry, if I've been kind of misleading when calling them arrays, but everyone I know uses that as it's name.

Reply:
I see, that clarifies it conceptually for me though, thanks! This won't be that much of a problem gladly, because we are just removing elements from the end of the array throughout this script. Thanks for the clarification.